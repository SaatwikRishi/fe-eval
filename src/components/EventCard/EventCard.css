.home-card {
  display: flex;
  flex-direction: column;
  height: inherit;
  background-color: white;
  width: inherit;
  justify-content: center;
  align-items: center;
  border-radius: 20px;
  margin: 10px;
  padding-bottom: 10px;
}
.home-card > img {
  border-radius: 10px 10px 0 0;
  height: 45%;
  width: 100%;
  margin-bottom: 1px;
}

.home-card-info {
  padding: 0 20px;
  display: flex;
  flex-direction: column;
  justify-content: space-around;

  row-gap: 20px;
  height: 55%;
  width: 100%;
  background-color: purple;
  margin-top: 1px;
  text-align: center;
  border-radius: 0 0 10px 10px;
}

.basic-info {
  margin-top: 10px;
  color: white;
  display: flex;
  flex-direction: column;
  row-gap: 10px;
  height: 60%;
}

.basic-info > h3 {
  font-family: "Big Shoulders Text" !important;
  font-size: 20px;
  font-weight: bold;
}
.basic-info > p {
  font-size: medium;

  min-width: 100%;
  min-height: 14%;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

.basic-info > h5 {
  line-clamp: 4;
  white-space: pre-wrap;
  min-height: 38%;
  overflow: hidden;
  text-overflow: ellipsis;
}
