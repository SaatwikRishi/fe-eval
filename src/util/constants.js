export const BASE_URL = 'http://localhost:8000/api/'

export function GET_THEMES () {
  return {
    url: 'themes'
  }
}

export function GET_EVENTS () {
  return { url: 'events' }
}

export function GET_EVENT_BY_ID (id) {
  return { url: 'events/' + id }
}
export function REGISTER_FOR_EVENT (id, isRegistered) {
  return {
    url: 'events/' + id,
    method: 'PATCH',
    data: { isRegistered }
  }
}
export function BOOKMARK_EVENT (id, isBookmarked) {
  return {
    url: 'events/' + id,
    method: 'PATCH',
    data: { isBookmarked }
  }
}

export function UPDATE_THEME (preferredThemeId) {
  return {
    url: 'themes',
    method: 'PUT',
    data: { preferredThemeId }
  }
}
