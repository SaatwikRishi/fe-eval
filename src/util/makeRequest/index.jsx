import axios, { AxiosError } from 'axios'
import { BASE_URL } from '../constants'
export const makeRequest = async (config) => {
  try {
    const response = await axios({ ...config, baseURL: config.baseUrl ?? BASE_URL })
    return response.data
  } catch (error) {
    if (error instanceof AxiosError) {
      throw new Error(error)
    } else {
      throw new Error('Something went Wrong')
    }
  }
}
