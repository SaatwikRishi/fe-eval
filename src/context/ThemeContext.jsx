import React, { useEffect } from 'react'
import PropTypes from 'prop-types'
import { makeRequest } from '../util/makeRequest'
import { GET_THEMES } from '../util/constants'
import { useNavigate } from 'react-router-dom'

export const ThemeContext = React.createContext()

export default function ThemeState ({ children }) {
  const [themes, setThemes] = React.useState([])
  const [loading, setLoading] = React.useState(true)
  const [currentTheme, setCurrentTheme] = React.useState(1)
  const getCurrentTheme = () => {
    const th = themes.filter((e) => e.id === currentTheme)[0].colorHexCode
    return th
  }

  const navigate = useNavigate()

  useEffect(() => {
    makeRequest(GET_THEMES())
      .then((e) => {
        setCurrentTheme(e.preferredThemeId)

        setThemes(e.themes)
        setTimeout(() => {
          setLoading(false)
        })
      }, 300)
      .catch((e) => {
        navigate('/error/500')
        setLoading(false)
      })
  }, [])
  if (loading) return <div>Loading</div>
  return (
    <ThemeContext.Provider value={{ themes, setCurrentTheme, getCurrentTheme }}>
      {children}
    </ThemeContext.Provider>
  )
}

ThemeState.propTypes = {
  children: PropTypes.node.isRequired
}
