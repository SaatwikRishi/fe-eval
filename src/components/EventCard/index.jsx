import React, { useContext } from 'react'

import './EventCard.css'
import PropTypes from 'prop-types'
import EventCardRegistered from '../EventCardRegistered'
import CustomButton from '../CustomButton'
import { ThemeContext } from '../../context/ThemeContext'
import { makeRequest } from '../../util/makeRequest'
import { REGISTER_FOR_EVENT } from '../../util/constants'
export default function EventCard ({ event, showRegistration = false }) {
  const { getCurrentTheme } = useContext(ThemeContext)
  const eventDate = new Date(event.datetime)
  const [registered, setRegistered] = React.useState(event.isRegistered)

  const handleEventRegistration = async () => {
    makeRequest(REGISTER_FOR_EVENT(event.id, !registered)).then(res => {
      setRegistered(!registered)
    })
  }

  return (
    <div
      className="home-card"
      style={{
        // backgroundColor: getCurrentTheme()
      }}
    >
      <img src={event.imgUrl} />
      <div
        className="home-card-info"
        style={{
          backgroundColor: getCurrentTheme()
        }}
      >
        <div className="basic-info">
          <h3>{event.name}</h3>
          <h5>{event.description}</h5>
          <p>
            <strong>VENUE: </strong>
            {event.venue}
          </p>
          <p>
            <strong>DATE: </strong>
            {eventDate.toDateString('DD MM YYYY', { timeZone: event.timezone })}
          </p>
        </div>
        <EventCardRegistered bookmarked={event.isBookmarked} id={event.id} registered={registered} seatsAvailable={event.areSeatsAvailable} />
        {event.areSeatsAvailable === true && showRegistration && <div> <CustomButton text= {registered ? 'UNREGISTER' : 'REGISTER'} onClick={handleEventRegistration} /> </div>}

      </div>
    </div>
  )
}
EventCard.propTypes = {
  event: PropTypes.object.isRequired,
  showRegistration: PropTypes.object.isRequired
}
