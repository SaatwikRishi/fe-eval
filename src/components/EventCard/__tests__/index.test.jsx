import EventCard from '..'
import { ThemeContext } from '../../../context/ThemeContext'
import React from 'react'
import { fireEvent, render, screen, waitFor } from '@testing-library/react'
import { makeRequest } from '../../../util/makeRequest'

jest.mock('../../../util/makeRequest', () => ({
  makeRequest: jest.fn()
}))

describe('Tests for Event Card', () => {
  it('should render when component is called and show registration button is true', () => {
    const { container } = render(
      <ThemeContext.Provider
        value={{
          getCurrentTheme: jest.fn()
        }}
      >
        <EventCard
          event={{
            id: 1,
            name: 'Battle of the Bands',
            description:
              "Get ready for Battle of the Bands, where the hottest up-and-coming rock groups will compete for the ultimate prize. With heart-pumping beats and electrifying performances, you won't want to miss this adrenaline-fueled event. Each band will bring their A-game, leaving everything on the stage for the chance to be crowned champion. Don't miss out on the most unforgettable rock competition of the year!",
            venue: 'All Stars Arena, Las Vegas, NV, USA',
            datetime: '2023-03-01T05:00:00.000Z',
            timezone: 'America/Los_Angeles',
            areSeatsAvailable: true,
            isRegistered: false,
            isBookmarked: false,
            imgUrl: 'https://i.ibb.co/3zbdvWX/battle-of-bands.jpg'
          }}
          showRegistration={true}
        />
      </ThemeContext.Provider>
    )
    expect(container).toMatchSnapshot()
  })
  it('should render when component is called and show registration button is false', () => {
    const { container } = render(
      <ThemeContext.Provider
        value={{
          getCurrentTheme: jest.fn()
        }}
      >
        <EventCard
          event={{
            id: 1,
            name: 'Battle of the Bands',
            description:
              "Get ready for Battle of the Bands, where the hottest up-and-coming rock groups will compete for the ultimate prize. With heart-pumping beats and electrifying performances, you won't want to miss this adrenaline-fueled event. Each band will bring their A-game, leaving everything on the stage for the chance to be crowned champion. Don't miss out on the most unforgettable rock competition of the year!",
            venue: 'All Stars Arena, Las Vegas, NV, USA',
            datetime: '2023-03-01T05:00:00.000Z',
            timezone: 'America/Los_Angeles',
            areSeatsAvailable: true,
            isRegistered: false,
            isBookmarked: false,
            imgUrl: 'https://i.ibb.co/3zbdvWX/battle-of-bands.jpg'
          }}
          showRegistration={false}
        />
      </ThemeContext.Provider>
    )
    expect(container).toMatchSnapshot()
  })

  it('should call handleEventRegistration when button is called', async () => {
    makeRequest.mockResolvedValue({ data: {} })
    const { container } = render(
      <ThemeContext.Provider
        value={{
          getCurrentTheme: jest.fn()
        }}
      >
        <EventCard
          event={{
            id: 1,
            name: 'Battle of the Bands',
            description:
              "Get ready for Battle of the Bands, where the hottest up-and-coming rock groups will compete for the ultimate prize. With heart-pumping beats and electrifying performances, you won't want to miss this adrenaline-fueled event. Each band will bring their A-game, leaving everything on the stage for the chance to be crowned champion. Don't miss out on the most unforgettable rock competition of the year!",
            venue: 'All Stars Arena, Las Vegas, NV, USA',
            datetime: '2023-03-01T05:00:00.000Z',
            timezone: 'America/Los_Angeles',
            areSeatsAvailable: true,
            isRegistered: false,
            isBookmarked: false,
            imgUrl: 'https://i.ibb.co/3zbdvWX/battle-of-bands.jpg'
          }}
          showRegistration={true}
        />
      </ThemeContext.Provider>
    )
    fireEvent.click(screen.getByTestId('btn')
    )
    await waitFor(() => {
      expect(screen.getByText('REGISTERED')).toBeTruthy()
    })
  })
})
