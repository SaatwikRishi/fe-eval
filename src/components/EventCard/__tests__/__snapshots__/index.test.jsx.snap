// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for Event Card should render when component is called 1`] = `
<div>
  <div
    class="home-card"
  >
    <img
      src="https://i.ibb.co/3zbdvWX/battle-of-bands.jpg"
    />
    <div
      class="home-card-info"
    >
      <div
        class="basic-info"
      >
        <h3>
          Battle of the Bands
        </h3>
        <h5>
          Get ready for Battle of the Bands, where the hottest up-and-coming rock groups will compete for the ultimate prize. With heart-pumping beats and electrifying performances, you won't want to miss this adrenaline-fueled event. Each band will bring their A-game, leaving everything on the stage for the chance to be crowned champion. Don't miss out on the most unforgettable rock competition of the year!
        </h5>
        <p>
          <strong>
            VENUE: 
          </strong>
          All Stars Arena, Las Vegas, NV, USA
        </p>
        <p>
          <strong>
            DATE: 
          </strong>
          Wed Mar 01 2023
        </p>
      </div>
      <div
        class="home-card-action"
      >
        <span
          class="action-registered"
        />
        <span
          class="action-bookmarked"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-bookmark fa-xl "
            data-icon="bookmark"
            data-prefix="far"
            focusable="false"
            role="img"
            viewBox="0 0 384 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 48C0 21.5 21.5 0 48 0l0 48V441.4l130.1-92.9c8.3-6 19.6-6 27.9 0L336 441.4V48H48V0H336c26.5 0 48 21.5 48 48V488c0 9-5 17.2-13 21.3s-17.6 3.4-24.9-1.8L192 397.5 37.9 507.5c-7.3 5.2-16.9 5.9-24.9 1.8S0 497 0 488V48z"
              fill="currentColor"
            />
          </svg>
        </span>
      </div>
    </div>
  </div>
</div>
`;
